### PR TITLE
refactor(telemetry-generator): Use pnpm

### PR DESCRIPTION
**Pull request by alexvy86**
## Description

Telemetry-generator was the last bit of the repo still using `npm` instead of `pnpm`. This fixes that and updates all places where we use it to reflect the change.

Took the chance to update its dependencies to more recent versions, syncing some (e.g. typescript) with the version we use in the client release group.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

A few runs of a test/ branch with these changes (plus a small adjustment in the build - client pipeline so that it logged telemetry during that run; already reverted in the latest version of this PR) (all msft internal):
- https://dev.azure.com/fluidframework/internal/_build/results?buildId=312314&view=results
- https://dev.azure.com/fluidframework/internal/_build/results?buildId=312319&view=results
- https://dev.azure.com/fluidframework/internal/_build/results?buildId=312385&view=results
- https://dev.azure.com/fluidframework/internal/_build/results?buildId=312384&view=results
- https://dev.azure.com/fluidframework/internal/_build/results?buildId=312386&view=results
- https://dev.azure.com/fluidframework/internal/_build/results?buildId=312387&view=results

I confirmed the expected telemetry shows up in Kusto in all cases.

[AB#4243](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/4243)

### CODE REVIEW

File: tools/telemetry-generator/src/handlers/testPassRate.ts [hunk 1]
```diff
@@ -11,7 +11,6 @@
  * @param logger - An ITelemetryBufferedLogger. Call its send() method to write the output telemetry events.
  */
 module.exports = function handler(fileData, logger): void {
-	// eslint-disable-next-line @rushstack/no-new-null -- External API can return an actual null
```
Comment (alexvy86) [inline tools/telemetry-generator/src/handlers/testPassRate.ts:14]: Eslint isn't complaining about this anymore 🤔

File: tools/telemetry-generator/.eslintrc.cjs [hunk 2]
```diff
@@ -21,4 +21,16 @@ module.exports = {
...
+			// Rules only for test files
+			files: ["*.spec.ts", "src/test/**"],
+			rules: {
+				// Disabling for now
+				"@typescript-eslint/no-require-imports": "off",
+				"@typescript-eslint/no-var-requires": "off",
+				"import/no-internal-modules": "off",
+			},
+		},
+	],
```
Comment (alexvy86) [inline tools/telemetry-generator/.eslintrc.cjs:35]: All of these things complain today in main. Not sure if we can make the handlers ESM, so until we investigate that, just disabling the rules.

Comment (Josmithr) [reply tools/telemetry-generator/.eslintrc.cjs:35]: Can we file a work item and reference it in a TODO comment?

File: tools/telemetry-generator/package.json [hunk 3]
```diff
@@ -35,21 +35,22 @@
 		"applicationinsights": "^2.4.1"
 	},
 	"devDependencies": {
-		"@fluid-internal/mocha-test-setup": "^2.3.1",
-		"@fluidframework/build-common": "^2.0.1",
-		"@fluidframework/eslint-config-fluid": "^5.2.0",
+		"@fluid-internal/mocha-test-setup": "^2.11.0",
```
Comment (Josmithr) [inline tools/telemetry-generator/package.json:38]: Nit: would it be worth updating these deps in a separate PR?

**Review by Josmithr — APPROVED** [review]

File: tools/telemetry-generator/src/handlers/testPassRate.ts (hunk 1 above)
Comment (Josmithr) [reply tools/telemetry-generator/src/handlers/testPassRate.ts:14]: I wonder if the rule was updated to allow comparisons to null? But I wouldn't be on that.

Comment (alexvy86) [reply tools/telemetry-generator/src/handlers/testPassRate.ts:14]: > The @rushstack/no-new-null rule flags type definitions with null that can be exported or used by others. The rule ignores declarations that are local variables, private members, or types that are not exported.

https://github.com/microsoft/rushstack/tree/main/eslint/eslint-plugin#rushstackno-new-null

I think it might have been added unnecessarily from the start. Adding a type that includes `null` does trigger the rule.

File: tools/telemetry-generator/.eslintrc.cjs (hunk 2 above)
Comment (alexvy86) [reply tools/telemetry-generator/.eslintrc.cjs]: Done.

File: tools/telemetry-generator/package.json (hunk 3 above)
Comment (alexvy86) [reply tools/telemetry-generator/package.json:38]: Since all the checks passed, I think there's little value in splitting it at this point, I'll keep them here.

File: tools/pipelines/templates/include-telemetry-setup.yml [hunk 4]
```diff
@@ -83,15 +87,17 @@ steps:
...
-    command: 'custom'
-    customCommand: 'install @ff-internal/aria-logger'
-    customRegistry: 'useNpmrc'
+    targetType: 'inline'
+    workingDirectory: ${{ parameters.pathToTelemetryGenerator }}
+    script: |
+      set -eu -o pipefail
+      # The telemetry-generator package is the root of its workspace so we do want to install this for the workspace
+      # root with '-w'. Without it, the install will complain and fail.
+      pnpm install -w @ff-internal/aria-logger;
```
Comment (jason-ha) [inline tools/pipelines/templates/include-telemetry-setup.yml:100]: We do a specific install of just `@ff-internal/aria-logger` and then later a full install. Why?

File: tools/pipelines/templates/include-telemetry-setup.yml [hunk 5]
```diff
@@ -101,5 +107,5 @@ steps:
     workingDirectory: ${{ parameters.pathToTelemetryGenerator }}
     script: |
       set -eu -o pipefail
-      npm i;
+      pnpm i;
```
Comment (jason-ha) [inline tools/pipelines/templates/include-telemetry-setup.yml:110]: Can we make install its own step?

File: tools/telemetry-generator/package.json (hunk 3 above)
Comment (jason-ha) [reply tools/telemetry-generator/package.json:38]: (It is fairly hard to migrate package managers and not get dep updates. So, I don't worry too much about getting more updates in one shot.)

File: tools/telemetry-generator/package.json [hunk 6]
```diff
@@ -35,21 +35,22 @@
...
 		"mocha": "^10.7.3",
 		"mocha-multi-reporters": "^1.5.1",
 		"prettier": "~3.0.3",
-		"rimraf": "^2.6.2",
+		"rimraf": "^5.0.10",
 		"sinon": "^19.0.2",
-		"typescript": "~4.5.5"
+		"typescript": "~5.4.5"
 	},
+	"packageManager": "pnpm@8.15.8+sha512.d1a029e1a447ad90bc96cd58b0fad486d2993d531856396f7babf2d83eb1823bb83c5a3d0fc18f675b2d10321d49eb161fece36fe8134aa5823ecd215feed392",
```
Comment (jason-ha) [inline tools/telemetry-generator/package.json:53]: That is awfully specific. Can you comment that somewhere?

File: tools/pipelines/templates/include-telemetry-setup.yml (hunk 4 above)
Comment (alexvy86) [reply tools/pipelines/templates/include-telemetry-setup.yml:100]: Because aria-logger is not part of the dependencies of the package; as an internal-only part of our infra, we're basically adding it to the package at this point during the pipeline run, and then the full install is to bring in all the deps that were already in package.json.

File: tools/telemetry-generator/pnpm-lock.yaml [hunk 7]
```diff
@@ -0,0 +1,3864 @@
...
+importers:
+
+  .:
+    dependencies:
+      '@fluidframework/core-interfaces':
+        specifier: ^2.0.0-rc.1.0.4
+        version: 2.11.0
+      '@fluidframework/test-driver-definitions':
+        specifier: ^2.0.0-rc.1.0.4
+        version: 2.0.0-rc.2.0.9
```
[the remark this answers is not in the document]
Comment (jason-ha) [reply tools/telemetry-generator/pnpm-lock.yaml:16]: Well, at least core-interfaces resolved to production. `test-driver-definitions` is now in `@fluid-internal` scope, which is why it got stuck on rc2.

[the remark this answers is not in the document]
Comment (alexvy86) [reply tools/telemetry-generator/pnpm-lock.yaml:16]: Almost certainly yes, I totally missed them. Updating now and I'll push the change shortly.

File: tools/pipelines/templates/include-telemetry-setup.yml (hunk 5 above)
Comment (alexvy86) [reply tools/pipelines/templates/include-telemetry-setup.yml:110]: Sure, I'll split it

File: tools/pipelines/templates/include-telemetry-setup.yml (hunk 4 above)
Comment (jason-ha) [reply tools/pipelines/templates/include-telemetry-setup.yml:100]: The pnpm command to add a dependency is `add`. I have only seen `install` used for installing existing dependencies. If this is an extra installation and not part of the regular deps, I would recommend it happen after regular install. Actually, I am surprised the regular install after this doesn't wipe this out.

File: tools/telemetry-generator/package.json (hunk 6 above)
Comment (jason-ha) [reply tools/telemetry-generator/package.json:53]: That comment.
I am surprised `pnpm@8.15.8` isn't enough to be version specific, but okay.
Do you know when packageManager versus engines matter or interact? We just list pnpm 8 under engines.

File: tools/pipelines/templates/include-telemetry-setup.yml (hunk 4 above)
Comment (jason-ha) [reply tools/pipelines/templates/include-telemetry-setup.yml:100]: And since there is not version specified here, I am assuming that we are always happy with it picking the latest it sees.
This feels like a toolchain/SBOM hole.

Comment (alexvy86) [reply tools/pipelines/templates/include-telemetry-setup.yml:100]: Yeah, it currently is. At some point it was a deliberate decision to let the version float so that improvements in aria-logger didn't require a separate change here to start having effect, but last time this topic came up in some private chat, I think we agreed pinning it would make sense, especially to prevent breaking older branches like LTS. That wouldn't prevent newer versions of transitive dependencies being brought in though, unless we first pin all those in aria-logger, which might be the right thing to do.

File: tools/telemetry-generator/package.json (hunk 6 above)
Comment (alexvy86) [reply tools/telemetry-generator/package.json:53]: I don't. Maybe @tylerbutler can comment when he's back. Or when he sees this comment while OOF and can't resist the temptation to reply like I've seen him do in the past😝.

File: tools/pipelines/templates/include-telemetry-setup.yml (hunk 4 above)
Comment (jason-ha) [reply tools/pipelines/templates/include-telemetry-setup.yml:100]: If we are going to pin, shouldn't we just make it part of regular dependencies and the lockfile?

File: tools/telemetry-generator/package.json (hunk 6 above)
Comment (jason-ha) [reply tools/telemetry-generator/package.json:53]: Just hopefully not while driving :)

File: tools/pipelines/templates/include-telemetry-setup.yml
Comment (jason-ha) [inline tools/pipelines/templates/include-telemetry-setup.yml:1]: In the .npmrc generation, should we add `frozen-lockfile=true` like we do for all our in-place .npmrc files?
I thought under CI that was supposed to be the default, but explicit is good.
I don't know what that means for one install of aria-logger.

**Comment by alexvy86** [comment]
Still looking into the option of further simplifying our use of telemetry-generator by making aria-logger an optional dependency, or maybe moving telemetry-generator to another repo entirely. But I'll go ahead and merge this as a net win even if ends up being a step towards the end state.
